easyfsl/methods/relation_networks.py

```diff
@@ -0,0 +1,177 @@
+"""
+See original implementation at
+https://github.com/floodsung/LearningToCompare_FSL
+"""
+
+import torch
+import torch.nn as nn
+from easyfsl.methods import AbstractMetaLearner
+from easyfsl.utils import compute_prototypes
+
+
+class RelationNetworks(AbstractMetaLearner):
+    """
+    Sung, Flood, Yongxin Yang, Li Zhang, Tao Xiang, Philip HS Torr, and Timothy M. Hospedales.
+    "Learning to compare: Relation network for few-shot learning." (2018)
+    https://openaccess.thecvf.com/content_cvpr_2018/papers/Sung_Learning_to_Compare_CVPR_2018_paper.pdf
+
+    In the Relation Networks algorithm, we first extract feature maps for both support and query
+    images. Then we compute the mean of support features for each class (called prototypes).
+    To predict the label of a query image, its feature map is concatenated with each class prototype
+    and fed into a relation module, i.e. a CNN that outputs a relation score. Finally, the
+    classification vector of the query is its relation score to each class prototype.
+
+    Note that for most other few-shot algorithms we talk about feature vectors, because for each
+    input image, the backbone outputs a 1-dim feature vector. Here we talk about feature maps,
+    because for each input image, the backbone outputs a "feature map" of shape
+    (n_channels, width, height). This raises different constraints on the architecture of the
+    backbone: while other algorithms require a "flatten" operation in the backbone, here "flatten"
+    operations are forbidden.
+    """
+
+    def __init__(self, *args, inner_relation_module_channels: int = 8):
+        """
+        Build Relation Networks by calling the constructor of AbstractMetaLearner.
+
+        Raises:
+            ValueError: if the backbone doesn't outputs feature maps, i.e. if its output for a
+            given image is not a tensor of shape (n_channels, width, height)
+        """
+        super().__init__(*args)
+
+        if len(self.backbone_output_shape) != 3:
+            raise ValueError(
+                "Illegal backbone for Relation Networks. Expected output for an image is a 3-dim "
+                "tensor of shape (n_channels, width, height)."
+            )
+
+        # Relation Networks use Mean Square Error.
+        # This is odd because this is a classification problem.
```

Comment:
C'est peut être un peu "méchant" comme tournure: While this is odd ..., the authors explains this choice by the fact ...  ?